explainer.md

```diff
@@ -36,193 +97,126 @@ self.addEventListener('launch', event => {
-  return a Boolean that indicates whether to cancel or default.
-* I would like to make sure that notifications can be shown from a `launch`
-  event handler. This satisfies use cases like a "save for later" tool that has
-  a share target. When the share target is chosen, it just shows a notification
-  "Saved for later", but doesn't actually spawn a browsing context.
-* It isn't sufficient to just cancel with `preventDefault`, because that
-  requires you to make the cancel decision synchronously. Since the Clients API
-  is asynchronous, you should be able to inspect the existing clients in a
-  promise before deciding whether to override the navigation. Thus the event
-  needs to be an `ExtendableEvent`.
-* LaunchEvent could have a url member instead of a request member. However, the
-  handler would then have no access to the form data of POST requests.
-
-## Firing the `launch` event
-
-When the user agent [navigates](https://html.spec.whatwg.org/#navigate) a
-browsing context *browsingContext* to a URL *resource*, if *resource* is in the
-scope of an active service worker, the user agent **MAY** go through the
-`launch` flow instead of the normal navigation behaviour.
-
-(NOTE: I don't understand the details of how new browsing contexts are created
-when using `window.open` or "Open in new window". I'm just going to assume that
-the *browsingContext* is null for navigations that open a new window or tab, and
-figure out the "correct" details later.)
-
-The purpose of this "**MAY**" is to allow the user agent to determine in what
-situations to allow the target site to interfere with the navigation. It
-**SHOULD** generally be done when navigating from a resource outside the target
-service worker scope, but **SHOULD NOT** be done when navigating within the same
-service worker scope. It **SHOULD NOT** be done when the user has explicitly
-indicated the intended browsing context for the navigation (e.g., "Open in new
-tab" or "Open in new window").
-
-If the user agent opts to use the `launch` flow, it should do the following
-instead of the normal navigation algorithm, within the context of *resource*'s
-active service worker:
-
-1. Create a new `LaunchEvent` *event* with `type` set to `"launch"`, `url` set
-   to *resource* and `clientId` set to the service worker client ID of
-   *browsingContext*. (`clientId` is `null` if this navigation would open a new
-   browsing context.)
-2. Fire *event* at the service worker global scope. The event listener is
-   **triggered by user activation**.
-3. If *event* is **cancelled**, abort these steps (and do not continue with the
-   navigation).
-4. If *event*'s [[*handlePending*]] slot is null, abort these steps and proceed
-   with the normal navigation algorithm.
+Alternatively, rather than specifying a `type` attribute and class hierarchy, different types of event could be fired in different cases, such as `launch-navigation` and `launch-file`.
```

Comment:
Or we could have one event with a number of optional params /shrug/

Reply:
Good thought.